### PR TITLE
add TimelineAction

### PR DESCRIPTION
**Pull request by 2youyou2**
TimelineAction is used for decode exported data from CocoStudio mac version.
It's a better solution for Armature.

TimelineAction inherits from cocos2d:;Action, so it's more closed to cocos2d.
TimelineAction's struct is more flexible and clear than Armature.
Frame -> Timeline -> TimelineAction
Frame is easy to extend its type, you just need to inherit from Frame, and implement Frame::onEnter, Frame::apply.

TimelineActionCache will load data from json resource and cache it.
When creates a new TimelineAction, TimelineActionCache will clone one from cache.

NodeReader will create a node tree from json resourece.

To run a TimelineAction, you need to do like this: 

Node\* node = NodeReader::getInstance()->createNode("TimelineAction/boy_1.ExportJson");
TimelineAction\* action = TimelineActionCache::getInstance()->createAction("TimelineAction/boy_1.ExportJson");

node->runAction(action);
action->gotoFrameAndPlay(0, 60, true);

### CODE REVIEW

**Comment by minggo** [comment]
Two questions:
- Is it possible merge timeline action into engine Action system?
- Is it possible to reuse existing images?

**Comment by 2youyou2** [comment]
1.TimelineAction contains both NodeReader and TimelineAction, not a single Action. So if we want to merge timeline action into Action system, do we need to separate them.
1. It's possible to reuse existing images. 
   Each Timeline have an action tag, and NodeReader will create an TimelineActionData for each loaded Node as it's UserObject. 
   When node run a timeline action, timeline action will search node's all descendants.
   If descendants have a TimelineActionData, and could find Timelines have the same action tag with this TimelineActionData's action tag, then these Timelines will do action to the nodes.
   So if existing images have TimelineActionData as it's UserObject with right action tag, it's possible to reuse them

**Comment by minggo** [comment]
Separate action from reader is a better design. Then this action can be reused.

**Comment by 2youyou2** [comment]
The reader is already seperated now.
We have two readers, NodeReader for node, TimelineActionCache for TimelineAction.
Because TimelineActionCache have the capacity to cache TimelineAction, so it named cache instead of   reader

**Comment by minggo** [comment]
So is it possible to separate action from reader?
If it is possible, please modify the codes.

Is it possible to reuse images?
If it is possible, please modify it.

Thanks.

File: cocos/2d/CCTimelineAction.h [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+#include "2d/CCNode.h"
+#include "2d/CCAction.h"
+#include "2d/CCTimeLine.h"
+#include "base/CCRef.h"
+#include "renderer/CCRenderer.h"
+
+NS_CC_BEGIN
+
+typedef void (Ref::*SEL_FrameEventCallFunc)(Frame *);
+#define frameEvent_selector(_SELECTOR) (SEL_FrameEventCallFunc)(&_SELECTOR)
```
Comment (minggo) [inline cocos/2d/CCTimelineAction.h]: Please use std::function<> instead.

File: cocos/2d/CCTimelineAction.h [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    virtual bool isPlaying();
+
+    /** Set the animation speed, this will speed up or slow down the speed. */
+    virtual void  setTimeSpeed(float speed);
+    /** Get current animation speed. */
+    virtual float getTimeSpeed();
+
+    /** duration of the whole action*/
+    virtual void setDuration(int duration) { _duration = duration; }
+    virtual int  getDuration() { return _duration; }
```
Comment (minggo) [inline cocos/2d/CCTimelineAction.h]: virtual int  getDuration() -> virtual int  getDuration() const.

File: cocos/2d/CCFrame.h [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+    static TextureFrame* create();
+
+    TextureFrame();
+
+    virtual void setNode(Node* node);
+
+    virtual void onEnter(Frame *nextFrame) override;
+    virtual Frame* clone() override;
+
+    inline void setTexture(std::string texture) { _texture = texture;}
```
Comment (minggo) [inline cocos/2d/CCFrame.h]: setTexture  -> setTextureName?

File: cocos/editor-support/cocostudio/ActionTimeline/CCActionTimelineCache.cpp [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+    ZOrderFrame* frame = ZOrderFrame::create();
+
+    int zorder = DICTOOL->getIntValue_json(json, Value);
+    frame->setZOrder(zorder);
+
+    return frame;
+}
+
+}
+}
```
Comment (minggo) [inline cocos/editor-support/cocostudio/ActionTimeline/CCActionTimelineCache.cpp:374]: Please add a new line at the end of file.

File: tests/cpp-tests/Classes/ExtensionsTest/CocoStudioActionTimelineTest/ActionTimelineTestScene.h [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+};
+
+class TestTimelinePerformance : public ActionTimelineTestLayer
+{
+public:
+    virtual void onEnter();
+    virtual std::string title() const override;
+};
+
+#endif  // __ANIMATION_SCENE_H__
```
Comment (minggo) [inline tests/cpp-tests/Classes/ExtensionsTest/CocoStudioActionTimelineTest/ActionTimelineTestScene.h:88]: Please add a new line at the end of file.

**Pull request merged**